tentacle/src/transports/mod.rs

```diff
@@ -28,6 +36,40 @@ pub trait Transport {
     fn dial(self, address: Multiaddr) -> Result<Self::DialFuture>;
 }
 
+pin_project! {
+    pub struct TransportFuture<T, F> {
+        #[pin]
+        executed: T,
+        maker: PhantomData<F>
+    }
+}
+
+impl<T, F> TransportFuture<T, F>
+where
+    T: Future<Output = Result<(Multiaddr, F)>> + 'static,
+{
+    pub fn new(executed: T) -> TransportFuture<T, F> {
+        TransportFuture {
+            executed,
+            maker: PhantomData::default(),
+        }
+    }
+}
+
+unsafe impl<T, F> Send for TransportFuture<T, F> where T: Send {}
+
+impl<T, F> Future for TransportFuture<T, F>
+where
+    T: Future<Output = Result<(Multiaddr, F)>> + 'static,
+{
+    type Output = Result<(Multiaddr, F)>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
```

Comment:
I suggest to use unsafe code here if the `pin_project` macro was used only to pin the inner field:
```
        let executed = unsafe {
            let this = self.get_unchecked_mut();
            Pin::new_unchecked(&mut this.executed)
        };
        executed.poll(cx)
```

Reply:
The only reason I use `pin_project_lite` is that most of the libraries in the entire ecosystem have this dependency so that there will be no additional burden on the introduction. But it is also possible to write directly using `unsafe`